chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/CertifiedProductController.java

```diff
@@ -766,41 +766,63 @@ public void streamSEDDetailsDocumentContents(HttpServletResponse response)
             throws InvalidArgumentsException, ValidationException, EntityCreationException, EntityRetrievalException, ObjectMissingValidationException,
             IOException {
-
-        String acbIdStr = request.getPendingListing().getCertifyingBody().get(CertifiedProductSearchDetails.ACB_ID_KEY).toString();
-        if (StringUtils.isEmpty(acbIdStr)) {
-            throw new InvalidArgumentsException("An ACB ID must be supplied in the request body");
+        Long acbId = getAcbIdFromPendingListing(request.getPendingListing());
+        if (acbId == null) {
+            throw new InvalidArgumentsException(msgUtil.getMessage("pendlingListing.missingAcb"));
         }
-        Long acbId = Long.valueOf(acbIdStr);
-        if (pcpManager.isPendingListingAvailableForUpdate(acbId, request.getPendingListing().getId())) {
-            PendingCertifiedProductDTO pcpDto = new PendingCertifiedProductDTO(request.getPendingListing());
-            PendingValidator validator = validatorFactory.getValidator(pcpDto);
-            if (validator != null) {
-                validator.validate(pcpDto, false);
-            }
-            if (pcpDto.getErrorMessages() != null && pcpDto.getErrorMessages().size() > 0
-                    || (pcpDto.getWarningMessages() != null && pcpDto.getWarningMessages().size() > 0
-                    && !request.isAcknowledgeWarnings())) {
-                throw new ValidationException(pcpDto.getErrorMessages(), pcpDto.getWarningMessages());
+        PendingCertifiedProductDTO pcpDto = new PendingCertifiedProductDTO(request.getPendingListing());
+        validate(pcpDto, request.isAcknowledgeWarnings());
+        ResponseEntity<CertifiedProductSearchDetails> response = null;
+        boolean wasAvailable = pcpManager.markAsProcessingIfAvailable(acbId, pcpDto.getId());
+        if (wasAvailable) {
+            CertifiedProductSearchDetails createdListing = null;
+            try {
+                CertifiedProductDTO createdProduct = cpManager.createFromPending(pcpDto, request.isAcknowledgeWarnings());
+                createdListing = cpdManager.getCertifiedProductDetails(createdProduct.getId());
+                activityManager.addActivity(ActivityConcept.CERTIFIED_PRODUCT, createdListing.getId(),
+                        "Created a certified product", null, createdListing);
+            } catch (Exception ex) {
+                //TODO - alert on this message in datadog
+                LOGGER.error("Unexpected exception confirming pending listing " + pcpDto.getId() + ".", ex);
+                pcpManager.markAsNotProcessing(acbId, pcpDto.getId());
+            } finally {
+                response = getConfirmResponse(createdListing);
             }
+        } else {
+            throw new InvalidArgumentsException(msgUtil.getMessage("pendingListing.alreadyProcessing"));
+        }
+        return response;
+    }
```

Comment:
It might be a big lift, but would it be possible to get this logic out of the controller and into a manager?